README.md

```diff
@@ -1,16 +1,119 @@
+
+```ts
+import range, { SmartRange } from "@leoschur/smart-range";
+```
+
+### Params
+
+-   `start` {int} start of the range can be larger than end with negative step
+-   `end` {int} end of the range
+-   `step` {int} step between values in the range - can be negative defaults to 1 or -1 if end < start
+
+```ts
+const [start, end, step] = [0, 10, 2];
+const r1 = range(start, end, step);
+console.log([r1.start, r1.end, r1.step]); // => [0, 10, 2]
+const r2 = range(end, start);
+console.log([r2.start, r2.end, r2.step]); // => [10, 0, -1]
+```
+
+### Properties
+
+-   `.start` get/set start of the range, integer only
+-   `.end` get/set end of the range, integer only
+-   `.step` get/set step between values, integer only
+-   `.length` readonly returns resulting length of the range, computed on access
+-   `[index]` get number at index, only available when `SmartRange` is created with `range` function otherwise use `SmartRange.at(index)`
+
+### `SmartRange.next()` & `SmartRange.return()`
+
+Implements the [Iterator Protocol](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#the_iterator_protocol). The class uses a internal (private) counter that increases with each `next()` call. Changing one of the properties (`start`, `end`, `step`) after `next()` was already called will result in undefined behavior when `next()` is called again afterwards. The internal counter or rather Iterator can be reset by calling `return()`.
+
+### `SmartRange.map(callBack)`
+
+Applies callBack to each element and returns resulting Array. Same as `new Array(SmartRange, callBack)` or `[...SmartRange].map(callBack)`. Just for convenience.
+
+### `SmartRange.includes(number)` | `number in range(a,z)`
+
+Checks if number is included in the Range. The Proxy version does the same thing but it still offers default behavior when the number/ string passed is not convertible as number.
+
+```ts
+const r = range(5, 51, 5);
+console.log(r.includes(30)); // => true
+console.log(30 in r); // => true
+console.log(13 in r); // => false
+console.log("length" in r); // => true
+```
+
+If you need to check more numbers you should convert the SmartRange once to an Array and check on that because as of right now, its not checked by calculation but converted to an array and the call passed on.
+
+### `SmartRange.at(index)` | `(range(a,z))[index]`
+
+Returns value that would be at the index if the SmartRange were to be converted to an Array. The index can be negative as well to access values from the back.
+
+```ts
+const r = range(0, 10);
+console.log(r[-1] == r.at(r.length - 1)); // => true
+```
+
 ## TODO's
 
 -   Features
-    -   [ ] support Number
-    -   [ ] support BigInt
+    -   [ ] support floats (optional with precision loss)
+    -   [ ] support generator function as step?
     -   [ ] support chars (automatic conversion ASCII => charCode)
+    -   [ ] range collection
```

Comment:
Can you elaborate how you would expect slices to behave, maybe even with example? :)

Reply:
I thought about sth. like this:
https://www.programiz.com/python-programming/examples/list-slicing

So you could use the index operator like this:

```python
my_list = [1, 2, 3, 4, 5]

print(my_list[1:4:2])
```

Reply:
I don't quite see a usecase for this. Besides you could archive this already through using an other SmartRange as indicies.
This is kind of akward but here is what I mean:
```ts
const myList = range(1,6,1);
const indicies = range(1,4,2);
const result = [...(function*() { for (const i of indicies) yield myList[i];})()];
console.log(result); // => [2, 4] equivalent to myList[1:4:2]
```
Sadly the pythonic syntax is not possible in JS anyway, since JS doesn't allow multiple values for the square brackets. Converting a string would theoretically be possible but thats kind of akward as well (my_list[`${1}:${4}:${2}`]).